src/test/txrequest_tests.cpp

```diff
@@ -0,0 +1,539 @@
+            BOOST_CHECK_MESSAGE(ret == expected, "[" + comment + "] mismatching requestables");
+        });
+    }
+
+    uint256 NewTxHash(const std::vector<std::vector<uint64_t>>& orders = {})
+    {
+        uint256 ret;
+        bool ok;
+        do {
+            ret = InsecureRand256();
+            ok = true;
+            for (const auto& order : orders) {
+                for (size_t pos = 1; pos < order.size(); ++pos) {
+                    uint64_t prio_prev = m_runner.txrequest.ComputePriority(ret, order[pos - 1], true);
+                    uint64_t prio_cur = m_runner.txrequest.ComputePriority(ret, order[pos], true);
+                    if (prio_prev >= prio_cur) {
+                        ok = false;
+                        break;
+                    }
+                }
+                if (!ok) break;
+            }
+            if (ok) {
+                ok = m_runner.txhashset.insert(ret).second;
+            }
+        } while(!ok);
+        return ret;
+    }
+
+    GenTxid NewGTxid(const std::vector<std::vector<uint64_t>>& orders = {})
+    {
+        return {InsecureRandBool(), NewTxHash(orders)};
+    }
+
+    uint64_t NewPeer()
+    {
+        bool ok;
+        uint64_t ret;
+        do {
+            ret = InsecureRandBits(63);
+            ok = m_runner.peerset.insert(ret).second;
+        } while(!ok);
+        return ret;
+    }
+
+    std::chrono::microseconds Now() const { return m_now; }
+};
+
+/** Add to scenario a test with a single tx announced by a single peer.
+ *
+ * config is an integer between 0 and 31, which controls which variant of the test is used.
+ */
+void BuildSingleTest(Scenario& scenario, int config)
+{
+    auto peer = scenario.NewPeer();
+    auto gtxid = scenario.NewGTxid();
+    bool immediate = config & 1;
+    bool preferred = config & 2;
+    auto delay = immediate ? NO_TIME : RandomTime();
+
+    scenario.SetTestName(strprintf("Single(config=%i)", config));
+
+    // Receive an announcement, either immediately requestable or delayed.
+    scenario.ReceivedInv(peer, gtxid, preferred, immediate ? MIN_TIME : scenario.Now() + delay);
+    if (immediate) {
+        scenario.Check(peer, {gtxid}, 1, 0, 0, "s1");
+    } else {
+        scenario.Check(peer, {}, 1, 0, 0, "s2");
+        scenario.AdvanceTime(delay - MICROSECOND);
+        scenario.Check(peer, {}, 1, 0, 0, "s3");
+        scenario.AdvanceTime(MICROSECOND);
+        scenario.Check(peer, {gtxid}, 1, 0, 0, "s4");
+    }
+
+    if (config >> 3) { // We'll request the transaction
+        scenario.AdvanceTime(RandomTime());
+        auto expiry = RandomTime();
+        scenario.Check(peer, {gtxid}, 1, 0, 0, "s5");
+        scenario.RequestedTx(peer, gtxid, scenario.Now() + expiry);
+        scenario.Check(peer, {}, 0, 1, 0, "s6");
+
+        if ((config >> 3) == 1) { // The request will time out
```

Comment:
I think using a AND here will get you higher coverage of this case ?

Reply:
Not sure what you mean here. There are four paths through the cascade of `if`s here and each value of `config >> 3` selects one of them.